lib/scanner/walk.go

```diff
@@ -277,6 +244,67 @@ func (w *walker) walkAndHashFiles(ctx context.Context, fchan, dchan chan protoco
 
 		return err
 	}
+
+	return func(path string, info fs.FileInfo, err error) error {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		// Return value used when we are returning early and don't want to
+		// process the item. For directories, this means do-not-descend.
+		skip = nil
+		// info nil when error is not nil
+		if info != nil && info.IsDir() {
+			skip = fs.SkipDir
+		}
+
+		if err != nil {
+			l.Debugln("error:", path, info, err)
+			return skip
+		}
+
+		if path == "." {
+			return nil
+		}
+
+		if fs.IsInternal(path) {
+			l.Debugln("ignored (internal):", path)
+			return skip
+		}
+
+		if !utf8.ValidString(path) {
+			l.Warnf("File name %q is not in UTF8 encoding; skipping.", path)
+			return skip
+		}
+
+		if w.Matcher.Match(path).IsIgnored() {
+			// If the parent wasn't ignored already, set this path as the "highest" ignored parent
+			if ignoredParent == "" || !strings.HasPrefix(path, ignoredParent+string(fs.PathSeparator)) {
+				ignoredParent = path
+			}
+			l.Debugln("ignored (patterns):", path)
+			if !w.Matcher.SkipIgnoredDirs() {
+				return nil
+			}
+			return skip
+		}
+		if ignoredParent != "" {
+			// Add parent directories of the current, not ignored path if needed
+			if rel := strings.TrimPrefix(path, ignoredParent+string(fs.PathSeparator)); rel != path {
+				for _, name := range strings.Split(filepath.Base(rel), string(fs.PathSeparator)) {
+					if err = handleItem(ignoredParent); err != nil {
+						return err
+					}
+					ignoredParent = filepath.Join(ignoredParent, name)
```

Comment:
Like, how does this work. 

If I am inside a parent that is ignored (`a/b/c`), trim the prefix and get relative path to ignored parent (if I am `a/b/c/d` I get d), yet `path` is always the full path (`a/b/c/d`), so why are we comparing these?

Then we do some split on the filepath.Base, filepath base returns the last element after the last path separator, so what's the point of the split?

The we call ourselves with the parent?
Why can't this be handled by the walk itself, why do we have to help walk?

I am reading this but this makes no sense to me.

Reply:
I can totally relate to your reaction, now that I reread the code with a little distance... I shuffled the code around and added comments. Hopefully together with the following explanation of "why do it in the first place" it is more clear now.

We need to help walk in the following scenario: Unignored path below an ignored path, e.g. `foo/bar/baz` with
```
!baz
/foo
```
Going towards `foo/bar/baz`, walk tells us to handle `foo` and `foo/bar`, but we don't do anything with them, as they are ignored. Then we come upon `foo/bar/baz` and realize, that we need to add it to the database. However we also need to add the parent directories, as we ignored them before.